services/venice-router/src/main/java/com/linkedin/venice/router/api/VeniceDelegateMode.java

```diff
@@ -471,77 +471,108 @@ public <H, P extends ResourcePath<K>, K, R> Scatter<H, P, K> scatter(
+      try {
+        for (; currentPartition < partitionCount; currentPartition++) {
+          keysForCurrentPartition = keysPerPartition.get(currentPartition);
+          if (keysForCurrentPartition.isEmpty()) {
+            continue;
           }
-          keyPartitionSet.addKeyPartitions(entry.getValue(), partitionName);
-        } else {
-          try {
-            selectHostForPartition(
-                partitionName,
-                hosts,
-                entry.getValue(),
-                venicePath,
-                hostMap,
-                helixGroupNum,
-                assignedHelixGroupId);
-          } catch (RouterException e) {
-            /**
-             * We don't want to throw exception here to fail the whole request since for streaming, partial scatter is acceptable.
-             */
-            scatter.addOfflineRequest(
-                new ScatterGatherRequest<>(Collections.emptyList(), new TreeSet<>(entry.getValue()), partitionName));
+          List<Instance> hosts = veniceHostFinder.findHosts(
+              requestMethod,
+              resourceName,
+              venicePath.getStoreName(),
+              currentPartition,
+              veniceHostHealthMonitor);
+
+          if (hosts.isEmpty()) {
+            veniceScatter.addOfflineRequest(
+                new ScatterGatherRequest<>(Collections.emptyList(), new TreeSet<>(keysForCurrentPartition)));
+          } else if (hosts.size() == 1) {
+            Instance host = hosts.get(0);
+            KeyPartitionSet<Instance, RouterKey> keyPartitionSet =
+                hostMap.computeIfAbsent(host, k -> new KeyPartitionSet<>(Collections.singletonList(host)));
+            keyPartitionSet.addKeys(keysForCurrentPartition);
+          } else {
+            try {
+              selectHostForPartition(
+                  hosts,
+                  keysForCurrentPartition,
+                  venicePath,
+                  hostMap,
+                  helixGroupNum,
+                  assignedHelixGroupId);
+            } catch (RouterException e) {
+              /**
+               * We don't want to throw exception here to fail the whole request since for streaming, partial scatter is acceptable.
+               */
+              veniceScatter.addOfflineRequest(
+                  new ScatterGatherRequest<>(Collections.emptyList(), new TreeSet<>(keysForCurrentPartition)));
+            }
           }
+          // Important to clear the inner list since it is thread-local state, which will be re-used by the next request
+          keysForCurrentPartition.clear();
+        }
+      } finally {
+        for (; currentPartition < partitionCount; currentPartition++) {
+          // This would only happen if the body of the try threw an exception.
+          // If that were to happen we would do a final cleanup here out of an abundance of caution.
+          keysForCurrentPartition = keysPerPartition.get(currentPartition);
+          keysForCurrentPartition.clear();
```

Comment:
So the `inner` list will become larger if there are some large requests or some hot partitions.
And here we have logic to clear the list belonging to every partition, so the total complexity will be O(inner_list_size * partition_count) for cleanup.
Not sure whether this is a concern or not, but we should be aware of such side effect.

Reply:
Well `clear` only goes as far as the size of the list, even if its capacity is larger. So the time complexity is basically O(3*N) where N is the number of keys queried (once to populate lists, a second time to populate the treemaps, and a third time to clear). Having treemaps in the middle is kind of a PITA, maybe we could change that (with further refactorings in Alpini).

Reply:
I see. Essentially, it only clears the entries current request is using.